### PR TITLE
feat(server): Import data from Hardcover

### PR DESCRIPTION
**Pull request by JVT038**
## Description

Basically, like the title says, this PR adds the functionality to import book data from Hardcover. Any books that are in Grimmory's database will be updated according to their status in Hardcover. Their read status, rating, the last read time, and the finished date wil be updated in the database of Grimmory.

This makes it way easier for me to migrate to Grimmory, as (most) of my reading data is stored in Hardcover, and I'd like to move this data to Grimmory as well.

## Linked Issue

It's an issue from the Booklore repository.

## Changes

* Added new DTOs BookIdentifier and HardcoverBookProgress.
* Added four new methods in HardcoverSyncService: importHardcoverData, createNewProgressRecords, updateExistingProgress and parseHardcoverResponse
* importHardcoverData: This manages the overall import of Hardcover data and calls the other methods.
* createNewProgressRecords: If the key (user_id, book_id) doesn't exist yet in user_book_progress (but it does exist in book_metadata), then this method will insert the new rows into the user_book_progress table. It dynamically generates a native SQL query to insert multiple rows at once, instead of multiple insertion queries.
* updateExistingProgress: If the key (user_id, book_id) already exists in user_book_progress, then this method will go to the corresponding row in user_book_progress and update it according to the Hardcover data. PS: This method performs multiple UPDATE queries for each of the rows in user_book_progress, which might be bad for performance.
* parseHardcoverResponse: This method simply receives the response from the GraphQL API call and parses it into ArrayList<HardcoverBookProgress>, so that the other methods can properly interact with the Hardcover response.

## Manual Testing Steps

1. Add books to Hardcover
2. Set one or more books to a certain status (e.g. currently reading)
3. Leave one or more books to an unset status (so that it doesn't appear in user_book_progress)
4. Go to settings -> devices.
5. Add Hardcover API key.
6. Click on 'trigger import' button.
7. The data gets properly updated according to the hardcover data.

**Regression testing:**

I ran all the existing Java tests in HardcoverSyncServiceTest and they all passed. Furthermore, the Booklore data was still getting synced to Hardcover.

**Edge cases covered:**

I make sure to deal with null values (e.g. no reading date).

## Screenshots (Optional)

image

## Additional Context (Optional)

I made this PR back when Booklore was still a thing, and the maintainer seemed pretty receptive to this change, but it wasn't merge due to all the drama that took place.

Anyway, I updated my branch with the latest changes now.

The only things that lack right now (which is why it's marked as draft) are tests for the newly added methods. I'll add unit tests later, but I would first like to know what maintainers think of this PR before moving on to testing.

## AI Disclosure

None

## Checklist

- [ ] This PR links and implements an accepted issue.
- [x] This PR is a single focused change.
- [ ] There are new or updated tests validating this change.
- [x] I ran `just ui check` and `just api check`.
- [x] I have added screenshots if there were any UI changes.
- [x] I have disclosed any AI usage as per the organization AI Policy above.
- [x] I understand all of my submitted changes.

## Summary by CodeRabbit

* **New Features**
  * Added Hardcover import controls in device settings.
  * Users can import Hardcover reading data and choose whether to overwrite existing progress.
  * Imported data includes reading status, ratings, dates, editions, pages, and book identifiers.
  * Added success and failure notifications for import attempts.
* **Bug Fixes**
  * Prevented simultaneous Hardcover imports and provided feedback when imports are temporarily rate-limited.
* **Chores**
  * Environment files are now excluded from version control.

### CODE REVIEW

**Comment by zachyale** [comment]
@JVT038 I think this is a good feature to have and would be happy to make an associated issue for this. Would it make more sense to have a review step or detail view for importing when there are conflicts? What's your level of interest in implementing beyond what's in this current PR?

**Comment by JVT038** [comment]
Maybe a intermediary detail view for conflicts is a good idea, but I think that might be better for a separate PR. 

I would like to keep this PR smaller and just focused on the importing stuff. 

But I'm happy to know that you're open to merging this. Then I'll start working on (unit) tests.

**Comment by zachyale** [comment]
> Maybe a intermediary detail view for conflicts is a good idea, but I think that might be better for a separate PR.
> 
> I would like to keep this PR smaller and just focused on the importing stuff.
> 
> But I'm happy to know that you're open to merging this. Then I'll start working on (unit) tests.

Yes, definitely interested- sorry that it's taken so long for a maintainer to provide feedback here! Let me know when it's reviewable. When it is, I'll create a "Hardcover improvements" epic to encapsulate this work, adding an import/conflict view, per-book sync settings, etc.

**Comment by JVT038** [comment]
My apologies for the delay, but I have finally (sort of) gotten around to writing tests.
I mostly generated them with Claude and manually went through every single line of code to check its meaning and review the way it works. 
The reason I used Claude is because I honestly hate writing tests (it's dull and monotonous) and I wasn't really sure how to deal with mocking the database/api objects. So Claude really helped with understanding how to mock databases and APIs and stuff.

Once again, I have reviewed every line and additionally, all tests pass. 

Note: I haven't used any AI for the `HardcoverSyncService.java` file. I enjoyed working on the importing logic, so I did that manually (old-fashioned way). It's literally just the unit tests that are made with AI-assistance.

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java [hunk 1]
```diff
@@ -807,6 +893,177 @@ private Map<String, Object> executeGraphQL(GraphQLRequest request) {
         }
     }
 
+    /**
+     * Import all the data related to the books from an account from Hardcover
+     * @param userId The ID of the Booklore user
+     */
+    @Async
```
Comment (imnotjames) [inline backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java:900]: Instead of async and a custom endpoint for this, we have a tasks system.  Perhaps imports from external sources can fit in that?

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java [hunk 2]
```diff
@@ -807,6 +893,177 @@ private Map<String, Object> executeGraphQL(GraphQLRequest request) {
...
+            return hardcoverIds.get(existingBookProgress.getHardcoverBookId());
+        } else if (allIsbns10.containsKey(existingBookProgress.getIsbn10())) {
+            return allIsbns10.get(existingBookProgress.getIsbn10());
+        } else if (allIsbns13.containsKey(existingBookProgress.getIsbn13())) {
+            return allIsbns13.get(existingBookProgress.getIsbn13());
+        }
+        return null;
+    }
+
+    private @Nullable ArrayList<HardcoverBookProgress> parseHardcoverResponse(List<Map> user_books, Map<String, HardcoverBookProgress> allIsbns10, Map<String, HardcoverBookProgress> allIsbns13, Map<String, HardcoverBookProgress> hardcoverIds) throws
```
Comment (imnotjames) [inline backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java:1007]: Is this something that could be done with objectmapper / jackson?

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java [hunk 3]
```diff
@@ -514,13 +533,80 @@ query GetUserBookAndReads($bookId:Int!) {
...
+                    }
+                }
+                """, String.valueOf(user_books.size()), String.valueOf(HARDCOVER_LIMIT));
+            GraphQLRequest request = new GraphQLRequest();
+            request.setQuery(query);
+            Map<String, Object> response = executeGraphQL(request);
+            if (response == null) {
+                return null;
+            }
+            // Navigate the response to get book info
```
Comment (imnotjames) [inline backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java:571]: This parsing logic should probably all use an objectmapper - had you tried that?  Is there a reason why it wouldn't work?

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java [hunk 4]
```diff
@@ -807,6 +893,177 @@ private Map<String, Object> executeGraphQL(GraphQLRequest request) {
...
+            return;
+        }
+        String sql = "INSERT INTO `user_book_progress` (`user_id`, `book_id`, `last_read_time`, read_status, `date_finished`, `personal_rating`) VALUES (?, ?, ?, ?, ?, ?)";
+        List<BookIdentifier> toBeInsertedBooks = new ArrayList<>();
+        for (BookIdentifier bookWithNoProgress : noProgressBookIds) {
+            if (getHardcoverBook(allIsbns10, allIsbns13, hardcoverIds, bookWithNoProgress) != null) {
+                toBeInsertedBooks.add(bookWithNoProgress);
+            }
+        }
+        jdbcTemplate.batchUpdate(sql, toBeInsertedBooks, 100, (preparedStatement, bookIdentifier) -> {
```
Comment (imnotjames) [inline backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java:949]: Instead of direct jdbc we normally use repository patterns.

File: .gitignore [hunk 5]
```diff
@@ -5,6 +5,7 @@ build/
 !**/src/main/**/build/
 !**/src/test/**/build/
 example-chart/charts/
```
Comment (imnotjames) [inline .gitignore:7]: Unrelated change.

**Comment by JVT038** [comment]
Sorry for the long waiting time; every time that I wanted to look into developing this PR, I had to deal with lots of merge conflicts. 

Anyway, I'll try and address your comments somewhere in the near future.

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java (hunk 3 above)
Comment (JVT038) [reply backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java]: I did the parsing logic this way, because I was copying how the Hardcover API respones were parsed in the other methods.

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java (hunk 1 above)
Comment (JVT038) [reply backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java]: I started looking into this, but I found it difficult to work with the current state of the task system. Especially when I have to try and get the userId of the user that got the task, and there is currently no real, easy method to get the creater of the task.
I could obviously change the task system, but I don't know if I'm the right person to make (major) architectural changes to Grimmory.

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java (hunk 3 above)
Comment (JVT038) [reply backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java]: I fixed the parsing logic to use an objectmapper (the built-in body() method to convert to a class). https://docs.spring.io/spring-boot/reference/io/rest-client.html#io.rest-client.webclient

File: backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java (hunk 1 above)
Comment (JVT038) [reply backend/src/main/java/org/booklore/service/hardcover/HardcoverSyncService.java]: I have just pushed a commit that runs the Hardcover imports as a task. 

My main concern is that the Hardcover import task can now be executed from the task management panel.
image


I still gotta find a way to get rid of this.
